File: post/leetcode/python/binary-tree-maximum-node.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   justice_103
@Problem:  http://www.lintcode.com/problem/binary-tree-maximum-node
@Language: Python
@Datetime: 17-03-10 14:15
'''

class Solution:
    # @param {TreeNode} root the root of binary tree
    # @return {TreeNode} the max node
    def maxNode(self, root):
        # Write your code here
        if root is None:
            return None
        max_node = root
        if root.left is not None:
            max_l = self.maxNode(root.left)
            if max_node.val < max_l.val:
                max_node = max_l
        if root.right is not None:
            max_l = self.maxNode(root.right)
            if max_node.val < max_l.val:
                max_node = max_l
        return max_node
            
            

File: post/leetcode/cpp/binary-tree-maximum-node.cpp
/*
@Copyright:LintCode
@Author:   justice_103
@Problem:  http://www.lintcode.com/problem/binary-tree-maximum-node
@Language: C++
@Datetime: 17-03-10 14:15
*/

class Solution {
public:
    /**
     * @param root the root of binary tree
     * @return the max node
     */
    TreeNode* maxNode(TreeNode* root) {
        // Write your code here
        if (root == NULL) {
            return NULL;
        }
        auto max = root;
        if (root->left != NULL) {
            auto maxl = this->maxNode(root->left);
            max = max->val > maxl->val ? max : maxl;
        }
        if (root->right != NULL) {
            auto maxl = this->maxNode(root->right);
            max = max->val > maxl->val ? max : maxl;
        }
        
        return max;
    }
};